doc/developers/cython.rst

```diff
@@ -0,0 +1,158 @@
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+
+  .. code-block:: bash
+
+         print(f"{test_val=}")
+
+* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and there's `ongoing work to simplify
+  and unify that across the codebase <https://github.com/scikit-learn/scikit-learn/issues/25572>`_. For now, make sure you understand which concrete types are used ultimately.
+
+* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
+  (debug symbols and proper optimization). To create a new conda environment (which you might need to deactivate and
+  reactivate conda after building/installing) with a source-built CPython interpreter:
+
+  .. code-block:: bash
+
+         git clone https://github.com/python/cpython.git
+         conda create -n debug-scikit-dev
+         conda activate debug-scikit-dev
+         cd cpython
+         mkdir debug
+         cd debug
+         ../configure --prefix=$CONDA_PREFIX --with-pydebug
+         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
+         make install
+
+  * You might find this alias handy:
+
+    .. code-block::
+
+         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"
+
+  * which you can use with:
+
+    .. code-block::
+
+         cythonX --annotate source.pyx
+
+Tips for performances
+^^^^^^^^^^^^^^^^^^^^^
+
+* Understand the GIL in context for CPython (which problems it solves, what are its limitations) and get a good
+  understanding of when Cython will be mapped to C code free of interactions with CPython, when it will not, and when
+  it cannot (e.g. presence of interactions with Python objects, which include functions). In this regard,
+  `PEP073 <https://peps.python.org/pep-0703/>`_ provides a good overview and context and pathways for removal.
+
+* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
+
+* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryview are lightweight.
+
+* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and we better not use it IMO,
+  even if handling fewer dimensions in some context would be preferable.
+
+* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)
+
+* Inline methods and function when it makes sense
+
+* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.
+
+* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections for a line of
+  Cython code, the better" is a good rule of thumb.
+
+Cython Internals
```

Comment:
Is it possible to add some context to this section on why this might be relevant to devs of sklearn and possible best practices associated?

Reply:
This part better be introduced in Cython's documentation directly, yes.